### DISCUSSION TITLE
what should be written in the <> of getUserObject

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22079] [Q&A General] [answered]
Howdy,

I am reading a code as follows. My question is In the Material, why  eos_uo(getUserObject<**EOS1P**>("eos_uo")) but not  eos_uo(getUserObject<**EOS1P_IdealFluid**>("eos_uo"))? Is there any documentation related to this I can find in moose website?
I heard from others that the reason is that EOS1P_IdealFluid inherits from EOS1P, so if the parent object name is used, all the children objects can be used. We can also use the the child object name but only that child object can be used. Is this how it works in moose? Thank you.
```
--------------------------------------------------------------------------
Input File:
[UserObjects]
  [./eos]
    type = **EOS1P_IdealFluid**
    bulk_modulus = 2e+012
  [../]
[]
[Materials]
  [./area0]
    type = FluidWell_1p1c
    eos_uo = eos
  [../]
[]
----------------------------------------------------------------------------
Material:
FluidWell_1p1c::FluidWell_1p1c(const InputParameters & parameters)
  : FluidWellGeneral(parameters),
    eos_uo(getUserObject<**EOS1P**>("eos_uo")),
{
}
---------------------------------------------------------------------------
UserObject:
**EOS1P_IdealFluid**::EOS1P_IdealFluid(const InputParameters & parameters)
  : EOS1P(parameters),
{
}
---------------------------------------------------------------------------
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

Yes we use object oriented programming in MOOSE, it's the OO in MOOSE!
Basically if your class can work with a base/parent class, then it can work with all the derived/child classes.
Please read up on this programming concept, it ll help you understand MOOSE better

Guillaume

**Reply by Oops-Qiao** [reply]
Unbelievable smart! Thank you Guilaume.
